libs/nettrace/nettrace.go

```diff
@@ -0,0 +1,1068 @@
+package nettrace
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/google/gopacket/pcapgo"
+	"github.com/lithammer/shortuuid/v4"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+)
+
+// PacketCapture is a recording of all/some packets that arrived or left through
+// a given interface.
+// This is typically included alongside NetTrace and captured packets are filtered
+// to contain only those that correspond with the traced connections.
+type PacketCapture struct {
+	// InterfaceName : name of the interface on which the packets were captured
+	// (on either direction).
+	InterfaceName string
+	// SnapLen is the maximum number of bytes captured for each packet.
+	// Larger packets are (silently) returned truncated.
+	SnapLen uint32
+	// Packets : captured packets.
+	Packets []gopacket.Packet
+	// Truncated is returned as true if the capture does not contain all packets
+	// because the maximum allowed total size would be exceeded otherwise.
+	Truncated bool
+	// WithTCPPayload : true if packet capture was configured to include also
+	// TCP packets with non empty payload.
+	WithTCPPayload bool
+}
+
+// WriteTo writes packet capture to a file or a buffer or whatever w represents.
+func (pc PacketCapture) WriteTo(w io.Writer) error {
+	pw := pcapgo.NewWriter(w)
+	err := pw.WriteFileHeader(pc.SnapLen, layers.LinkTypeEthernet)
+	if err != nil {
+		return err
+	}
+	for _, packet := range pc.Packets {
+		err = pw.WritePacket(packet.Metadata().CaptureInfo, packet.Data())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// WriteToFile saves packet capture to a given file.
+func (pc PacketCapture) WriteToFile(filename string) error {
```

Comment:
Isn't io.Writer a general interface which can be used for writing to files? i.e. we can use WriteTo to write to io.Writer which is actually a file

Reply:
Yes, `WriteToFile` is just a simple helper for writing PCAP files that wraps `os.Create` and `pc.WriteTo` (to avoid users repeating the same code).